lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -703,46 +732,53 @@ class MongoClientInterface {
     putObject(bucketName, objName, objVal, params, log, cb) {
         MongoUtils.serialize(objVal);
         const c = this.getCollection(bucketName);
-        if (params && params.versioning && !params.versionId) {
-            return this.putObjectVerCase1(c, bucketName, objName, objVal,
-                params, log, cb);
-        } else if (params && params.versionId === '') {
-            return this.putObjectVerCase2(c, bucketName, objName, objVal,
-                params, log, cb);
-        } else if (params && params.versionId && !params.repairMaster) {
-            return this.putObjectVerCase3(c, bucketName, objName, objVal,
-                params, log, cb);
-        } else if (params && params.versionId && params.repairMaster) {
-            return this.putObjectVerCase4(c, bucketName, objName, objVal,
+        return this.getBucketVFormat(bucketName, log, (err, vFormat) => {
```

Comment:
Actually, I am wondering if this should be done in here.... It seems to me that reading the metastore to find the "type" of bucket is already done in Cloudserver : so IMHO it would be more consistant if it staid there only, and this function would just take the "vFormat" in argument...

Another argument for this approach is that cloudserver has likely alreday queried the metastore to retrieve information on the bucket, so this extra call to the backend here may actually be redundant in many/most case.

How is it implemented in other backeds? Do we have the same call to metastore from within the `putObject` method ?

Reply:
> How is it implemented in other backends?

Cloudserver doesn't have access to the vFormat attribut, the logic that deals with that is offloaded to the metadata backend, this is also the case for other backends. So Cloudserver does query the `metastore` to get bucket attributs but it never gets the vFormat.

> Do we have the same call to `metastore` from within the putObject method ?

In metadata it's a bit different, but as far as i understood yes we always query the db to get the vFormat at each request.

https://github.com/scality/MetaData/blob/613bc56a9f174dd4ba95f2e513dc648260bac3d9/lib/server/VersioningRequestProcessor.js#L407-L411
^ Here `replicateRequest` is used to replicate the putObject (and deleteObject) action among the servers. We get the vFormat each time we call it.

Reply:
In metadata the bucket metadata is cached once retrieved, which means that we only get them from the db when they're not in the cache.

Reply:
Do you recommend adding a cache here? and if so how do we decide its sizing ?